### PR TITLE
feat: add `Dockerfile`; add `docker-compose`

### LINKED ISSUE #202: Create docker(-compose) environment for ease of use purposes

### LINKED ISSUE #264: Remove SSL instructions from basic setup - assume users will use cloudflare (flex)
Doing this will remove complication from our setup - at the moment a significant amount of effort goes into creation of an ssl certificate - which is not required if using cloudflare's flex ssl option.

For simplicity, will be assuming users can use this by default - but we should also include ssl configuration documentation somewhere too.

### PR DESCRIPTION
**Pull request by 7mochi**
This is a continuation of #269. Closes #202, closes #264. Depends on #367, ~~depends on #411~~

- [x] Add Dockerfile
- [x] Add docker-compose
- [x] Update readme
- [x] Remove rust-cargo as dependency (requires wheels for [akatsuki-pp-py](https://github.com/osuAkatsuki/akatsuki-pp-py) before making changes)
- [x] Add nginx maxminddb

### CODE REVIEW

**Comment by cmyui** [comment]
@7mochi is this ready for review?

**Comment by 7mochi** [comment]
> @7mochi is this ready for review?

yep

**Comment by cmyui** [comment]
looks like nice work! sorry for delay as i'm a bit busy at the moment (akatsuki is rather popular at the moment & scaling it is taking most of my free time), so will need to wait a bit for a proper review

File: Dockerfile
Comment (aeongdesu) [inline Dockerfile:4]:
```suggestion
RUN apk add --no-cache \
```
consider reduce image size? https://stackoverflow.com/questions/49118579/alpine-dockerfile-advantages-of-no-cache-vs-rm-var-cache-apk

**Review by aeongdesu — CHANGES_REQUESTED** [review]

**Review by aeongdesu — APPROVED** [review]
LGTM
now add cloudflare flex SSL instructions 🙃

**Comment by NiceAesth** [comment]
Hey there, #332 may be of interest to you guys

**Comment by 7mochi** [comment]
I'll wait until #332 and #333 are merged to make the changes

**Comment by 7mochi** [comment]
> LGTM now add cloudflare flex SSL instructions upside_down_face

I just added the instructions, could you guys check it?

**Comment by aeongdesu** [comment]
> > LGTM now add cloudflare flex SSL instructions upside_down_face
> 
> I just added the instructions, could you guys check it?

Uploading images on github would be nice instead imgur

get url from this, they wont expire i think

Screenshot_20230108_125748_Kiwi Browser

**Comment by 7mochi** [comment]
> > > LGTM now add cloudflare flex SSL instructions upside_down_face
> > 
> > 
> > I just added the instructions, could you guys check it?
> 
> Uploading images on github would be nice instead imgur
> 
> get url from this, they wont expire i think
> 
> Screenshot_20230108_125748_Kiwi Browser

I thought there was no problem because the og README even has a discord image :tf:

**Comment by cmyui** [comment]
#339 will be merged very shortly

**Comment by cmyui** [comment]
it has now been merged

File: .env.example
Comment (cmyui) [inline .env.example:4]: if these are constant, no need to include them here

just the external host (which may vary) details are important

File: .env.example
Comment (cmyui) [inline .env.example:13]: generally just "mysql" and "redis" are standard for these, no need to prefix them imo

File: README.md
Comment (cmyui) [inline README.md:87]: i should be able to set this up. is it an issue of race conditions?

File: README.md
Comment (cmyui) [inline README.md:113]:
```suggestion

# Manual installation
```

File: README.md
Comment (cmyui) [inline README.md:159]: I think there's still value in keeping these instructions if users wish to do SSL themselves. Perhaps this could be split between a "cloudflare flex" setup, and a manual ssl setup?

Comment (7mochi) [reply README.md:159]: they are below

File: ext/nginx.conf.example
Comment (cmyui) [inline ext/nginx.conf.example:85]: this file is getting a bit complex, i wonder if we can make any simplifications.. might try this myself

File: main.py
Comment (cmyui) [inline main.py:44]: YAY this is gone 🎉

**Comment by cmyui** [comment]
overall pr looks good! very close to mergeable. @7mochi would you like to update to account for #339 or should i?

**Review by cmyui — CHANGES_REQUESTED** [review]
overall pr looks good! very close to mergeable. @7mochi would you like to update to account for #339 or should i?

**Comment by 7mochi** [comment]
> overall pr looks good! very close to mergeable. @7mochi would you like to update to account for #339 or should i?

could you do it? I'm pepega solving conflicts with git

File: .env.example
Comment (cmyui) [reply .env.example:4]: ah i see, it matters outside of docker setup..

File: .env.example
Comment (cmyui) [reply .env.example:13]: will be applied in next commit

File: README.md
Comment (7mochi) [reply README.md]: i think I've managed to fix it, check it when you are free

File: scripts/start_server.sh
Comment (NiceAesth) [inline scripts/start_server.sh:1]: If this script is intended to be used outside of docker as well, then I would suggest editing the makefile to run the script instead of having it the other way around since `ensure_connected_services` has been removed.

File: Dockerfile
Comment (NiceAesth) [inline Dockerfile:1]: Wouldn't `python3.9-slim` be a better choice than alpine linux? See: [why not to use alpine linux](https://megamorf.gitlab.io/2020/05/06/why-it-s-better-not-to-use-alpine-linux-for-python-projects)

File: Dockerfile
Comment (NiceAesth) [inline Dockerfile:31]: See comment on `start_server.sh`

File: README.md
Comment (NiceAesth) [inline README.md:270]: If I recall correctly, cloudflare provides a way to export/import DNS records. Could be a nice-to-have

File: README.md
Comment (NiceAesth) [inline README.md:303]: Would it not be possible to have these images in the repo itself? Perhaps in the `.github` directory

File: docker-compose.yml
Comment (NiceAesth) [inline docker-compose.yml:38]: `Pipenv` will load the `.env` file on its own. Is this really necessary?

Comment (7mochi) [reply docker-compose.yml:38]: leftovers from when pipenv wasn't used

File: README.md
Comment (7mochi) [reply README.md:303]: Wouldn't that increase the repo size? or it doesn't matter

File: scripts/start_server.sh
Comment (NiceAesth) [reply scripts/start_server.sh:1]: Hmm not too sure on this anymore since it relies on wait4x

Comment (cmyui) [reply scripts/start_server.sh:1]: on that topic - have any opinion on [wait4x](https://github.com/atkrad/wait4x) vs. [wait-for-it](https://github.com/vishnubob/wait-for-it)?
i've only used wait-for-it, it has ~20x more stars (so i assume more prod tested)

Comment (NiceAesth) [reply scripts/start_server.sh:1]: I would prefer wait-for-it myself for two main reasons:
1. wait4x seems to be fairly heavily focused on alpine linux (something which we don't want)
2. wait-for-it is a simple bash script which could easily be included (would solve the concern I had)

Comment (7mochi) [reply scripts/start_server.sh:1]: I saw that wait4x was in the [alpine repo](https://pkgs.alpinelinux.org/packages?name=wait4x&branch=edge&repo=&arch=&maintainer=) and decided to use it, but if we're going to switch to slim then let's use wait-for-it

Comment (NiceAesth) [reply scripts/start_server.sh:1]: wait-for-it is licensed under the MIT license, therefore distributing it ourselves is an option. @cmyui do you have any thoughts on this?

Comment (cmyui) [reply scripts/start_server.sh:1]: Yeah I've typically done that, [or have written by own lightweight version of it](https://github.com/akatsuki-v2/users-service/blob/main/scripts/await-service.sh) for my own usage.

Perhaps there's an argument to use something like a submodule to keep it up to date automatically, but I'll defer that choice to whoever implements it (seems rather minor imo).

**Comment by NiceAesth** [comment]
Will wait for wheels to be built for [akatsuki-pp-py](https://github.com/osuAkatsuki/akatsuki-pp-py) before making changes

- [ ] Remove rust and cargo as dependencies
- [ ] Switch to `python:3.9-slim`

**Comment by 7ez** [comment]
> Will wait for wheels to be built for [akatsuki-pp-py](https://github.com/osuAkatsuki/akatsuki-pp-py) before making changes
> 
>     * [ ]  Remove rust and cargo as dependencies
> 
>     * [ ]  Switch to `python:3.9-slim`

rust is needed to build pp system tho?

**Comment by cmyui** [comment]
> rust is needed to build pp system tho?

That's what the wheels fix - we build akatsuki-pp-rs on the side of the python bindings, so we don't need it here.

File: .env.example
Comment (NiceAesth) [reply .env.example:4]: can this be resolved?

File: Dockerfile
Comment (aeongdesu) [inline Dockerfile:7]:
```suggestion
    build-essential \
    && rm -rf /var/lib/apt/lists/*
```
https://askubuntu.com/a/1251250

**Review by aeongdesu — CHANGES_REQUESTED** [review]

**Comment by 7mochi** [comment]
Because #367 the nginx image has to be changed to [nginx-geoip2](https://hub.docker.com/r/anroe/nginx-geoip2)

**Comment by NiceAesth** [comment]
as a small disclaimer, you shouldn't mark this PR as ready for review until wheels are available

**Comment by NiceAesth** [comment]
use checklists from now on, no longer need to mark as draft

**Comment by NiceAesth** [comment]
i barely remember stuff from this PR anymore, lmao. @tsunyoku any progress on bindings?

**Comment by 7mochi** [comment]
> i barely remember stuff from this PR anymore, lmao. @tsunyoku any progress on bindings?

lmao, i'm still waiting 😭

**Comment by tsunyoku** [comment]
i was meant to move over to poetry so i could have the wheels but i forgot.. i will get to it tmrw. sorry guys!

**Comment by NiceAesth** [comment]
#411 ^ _ ^

File: Dockerfile
Comment (NiceAesth) [inline Dockerfile:10]: this can go bye bye

File: .gitignore
Comment (NiceAesth) [inline .gitignore:14]: any reason for this? i assume it's used by docker now (?)

File: Dockerfile
Comment (NiceAesth) [inline Dockerfile:4]: pin versions

File: Dockerfile
Comment (NiceAesth) [inline Dockerfile:13]: pin versions

File: README.md
Comment (NiceAesth) [reply README.md:270]: bumping this

File: README.md
Comment (NiceAesth) [inline README.md:301]: typo

File: README.md
Comment (NiceAesth) [inline README.md:316]: honestly i would much rather recommend full (strict) cloudflare setup with an origin certificate over generating your own

File: docker-compose.yml
Comment (NiceAesth) [inline docker-compose.yml:13]: data will not persist

File: docker-compose.yml
Comment (NiceAesth) [inline docker-compose.yml:25]: data will not persist

File: docker-compose.yml
Comment (NiceAesth) [inline docker-compose.yml:15]: defining ports like this is useless, by default all containers are in the same network in the same compose

File: Dockerfile
Comment (NiceAesth) [inline Dockerfile:14]: you are running as root, please create a new user and run under that
ex:
```docker
RUN useradd --create-home appuser
WORKDIR /home/appuser
USER appuser
```

File: docker-compose.yml
Comment (NiceAesth) [inline docker-compose.yml:39]: is this path correct? from what i can tell, you mounted the app to /prod

File: ext/nginx.conf.example
Comment (NiceAesth) [reply ext/nginx.conf.example:99]: bump for @cmyui

**Review by NiceAesth — CHANGES_REQUESTED** [review]

File: README.md
Comment (NiceAesth) [reply README.md:328]: i assume you didn't see this before the latest commit @7mochi
also http to https could be done via nginx

File: docker-compose.yml
Comment (NiceAesth) [reply docker-compose.yml:39]: as i said this path is incorrect

File: docker-compose.yml
Comment (7mochi) [reply docker-compose.yml:15]: it could be useful to access to the db from outside the container

File: docker-compose.yml
Comment (NiceAesth) [inline docker-compose.yml:18]: this expose directive does nothing

**Review by NiceAesth — APPROVED** [review]
i'm fine with the current state of the PR, was thinking about ipv6 but whatever, if you're using that you don't need a script to help you out

**Review by aeongdesu — APPROVED** [review]

**Review by cmyui — APPROVED** [review]
let's send it and see how it's accepted by the community

can make adjustments from there

**Pull request merged**
